Adding OAuth requirements draft v1

csfreak92 opened the pull request:
Adding OAuth requirements draft v1. Initial draft for voting how we shall present these OAuth requirements derived from the RFC standard in ASVS. 

<!---
IMPORTANT NOTES:
- Changes should always be made only in the raw .md files and not in the CSV, JSON, XLSX, PDF, DOCX files, etc.
- Please do not open a pull request without first opening an associated issue.
- Please carry out all discussion in the associated issue only.
- Please refer to the following link for guidance on labeling contributions https://github.com/OWASP/ASVS/blob/master/CONTRIBUTING.md#standard-for-changes
-->

This Pull Request relates to issue #996.

elarlang commented:
We should not do development in master branch.

I don't want this to get accidentally merged and then sort out the mess from the repo - can you cover all the proposed requirements in related (or new) issue?

One way how to do it - you open an issue, all the requirements are there and based on feedback you change the original post content till it's PR ready?

Or you do it in separate branch or in fork etc.

csfreak92 commented:
Woops, let me fix that. My bad I didn't see it. Thanks for double-checking!

elarlang commented:
@tghosth - maybe separate branch for developing this?

csfreak92 commented:
@tghosth, please let me know which branch shall I do a pull request for this ASVS requirements for OAuth.

tghosth commented:
I have changed the PR target and commented on the main thread https://github.com/OWASP/ASVS/issues/996#issuecomment-1374871231

jmanico reviewed (changes requested):
i would like to see more focus on sender constrained tokens per the BCP. If I steal your access or refresh token it should not be useful at all unless I also have the client secret per the new BCP.

SBVels commented:
Hi. Can try and thank you.S

danielcuthbert commented:
@csfreak92 just wanted to say thank you for creating these, they are brilliant!!

csfreak92 commented:
@tghosth, I'm having a hard time trimming the requirements down... I'll just push whatever I have now and let's figure out as a collective if there is anything else can be done. I don't like leaving out important details of the requirements and putting them in an explanatory section as I don't think a lot of folks read those parts anyways.

csfreak92 commented:
@tghosth, @jmanico, I pushed the finalized cleaned-up OAuth 2.0 requirements. This current commit still lives in Section 3.x - Session Management. I know we wish to move this to a new section on its own but I will let that numbering be on your hands as I don't want to mess any convention we have for the project. 

P.S. I don't know why the other commits from master got pulled here... I suppose that won't be an issue?

tghosth commented:
hi @csfreak92, note my comment here https://github.com/OWASP/ASVS/issues/996#issuecomment-1494623031

csfreak92 commented:
@tghosth, I just pushed a new commit trimming down the Authorization Server pieces from the requirements. Can you have a look and let me know if those are good and nothing is misleading in the language?

jmanico commented:
This all looks solid to me. @tghosth merge?

tghosth reviewed (changes requested):
Thanks @csfreak92, this is a tough area and I think this is much more focused than the original suggestion which is great.

I do have quite a few comments that hopefully you can go through.

At the end, we will also need to clean up the PR because it seems to have some unrelated changes in here as well...

csfreak92 commented on 5.0/en/0x12-V3-Session-management.md:
```diff
@@ -91,10 +90,88 @@ Attackers start a half-open attack by attempting to lock, reset, or recover a cr
+
+| # | Description | L1 | L2 | L3 |
+| :---: | :--- | :---: | :---: | :---: |
+| **3.8.1** | [ADDED] Verify that Clients do not expose URLs that forward the user's browser to arbitrary URIs obtained from a query parameter ("open redirectors") which can enable exfiltration of authorization codes and access tokens.  | ✓ | ✓ | ✓ |
```
I think you are right, this is somehow covered in 5.1.5, but we are just emphasizing here for OAuth-related security which I think makes it better that it mentions to prevent exfiltration of authorization codes and access tokens so that developers/engineers understand why it is important not to use URLs. I think we can keep this unless you strongly think otherwise?

tghosth replied:
I think it is already covered, I don't really want to double up as there are already so many extra requirements. I think let's remove it. Maybe we should have a note at the start of other important requirements that are not OAuth specific but often feature in OAuth related vulnerabilities?

csfreak92 replied:
Ah good point, I'll just add that OAuth specific info in 5.1.5 as a note, probably in the intro paragraph of that subsection.

tghosth replied:
Ok thanks

csfreak92 commented on 5.0/en/0x12-V3-Session-management.md:
```diff
@@ -91,10 +90,88 @@ Attackers start a half-open attack by attempting to lock, reset, or recover a cr
+| # | Description | L1 | L2 | L3 |
+| :---: | :--- | :---: | :---: | :---: |
+| **3.8.1** | [ADDED] Verify that Clients do not expose URLs that forward the user's browser to arbitrary URIs obtained from a query parameter ("open redirectors") which can enable exfiltration of authorization codes and access tokens.  | ✓ | ✓ | ✓ |
+| **3.8.2** | [ADDED] Verify Clients have ensured that the Authorization Server supports PKCE may rely on the CRSF protection provided by PKCE. In OpenID Connect flows, the "nonce" parameter provides CSRF protection. Otherwise, one-time user CSRF tokens carried in the "state" parameter that are securely bound to the user agent must be used for CSRF protection.  | ✓ | ✓ | ✓ |
```
I like that. I'll reword this then to reflect it. Quick question: the other sentences related to 3.8.2 here seems very useful info to anyone reading it too. Maybe I should add this on the header intro paragraph of the subsection for these requirements? What do you think?

tghosth replied:
Yeah add it there for now.

csfreak92 commented on 5.0/en/0x12-V3-Session-management.md:
```diff
@@ -91,10 +90,88 @@ Attackers start a half-open attack by attempting to lock, reset, or recover a cr
+| :---: | :--- | :---: | :---: | :---: |
+| **3.8.1** | [ADDED] Verify that Clients do not expose URLs that forward the user's browser to arbitrary URIs obtained from a query parameter ("open redirectors") which can enable exfiltration of authorization codes and access tokens.  | ✓ | ✓ | ✓ |
+| **3.8.2** | [ADDED] Verify Clients have ensured that the Authorization Server supports PKCE may rely on the CRSF protection provided by PKCE. In OpenID Connect flows, the "nonce" parameter provides CSRF protection. Otherwise, one-time user CSRF tokens carried in the "state" parameter that are securely bound to the user agent must be used for CSRF protection.  | ✓ | ✓ | ✓ |
+| **3.8.3** | [ADDED] Verify that when an OAuth Client can interact with more than one Authorization Server, Clients should use the issuer "iss" parameter as a countermeasure, or alternatively based on an "iss" value in the authorization response (such as the "iss" Claim in the ID Token in OpenID) | ✓ | ✓ | ✓ |
```
Let me double check somewhere I read it. I'll get back to you on this.

csfreak92 replied:
@tghosth, yes there is documentation on the "iss" parameter as expected. FYI, tagging them here below: 

[RFC9207] [I-D.ietf-oauth-iss-auth-resp]. https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics-18#section-2.1-4

and 

[I-D.ietf-oauth-iss-auth-resp]. https://datatracker.ietf.org/doc/html/draft-ietf-oauth-iss-auth-resp-00

csfreak92 commented on 5.0/en/0x12-V3-Session-management.md:
```diff
@@ -91,10 +90,88 @@ Attackers start a half-open attack by attempting to lock, reset, or recover a cr
+| **3.8.1** | [ADDED] Verify that Clients do not expose URLs that forward the user's browser to arbitrary URIs obtained from a query parameter ("open redirectors") which can enable exfiltration of authorization codes and access tokens.  | ✓ | ✓ | ✓ |
+| **3.8.2** | [ADDED] Verify Clients have ensured that the Authorization Server supports PKCE may rely on the CRSF protection provided by PKCE. In OpenID Connect flows, the "nonce" parameter provides CSRF protection. Otherwise, one-time user CSRF tokens carried in the "state" parameter that are securely bound to the user agent must be used for CSRF protection.  | ✓ | ✓ | ✓ |
+| **3.8.3** | [ADDED] Verify that when an OAuth Client can interact with more than one Authorization Server, Clients should use the issuer "iss" parameter as a countermeasure, or alternatively based on an "iss" value in the authorization response (such as the "iss" Claim in the ID Token in OpenID) | ✓ | ✓ | ✓ |
+| **3.8.4** | [ADDED] Verify that when the other countermeasure options for OAuth clients interacting with more than one Authorization Servers are absent, Clients may instead use distinct redirect URIs to identify authorization endpoints and token endpoints.  | ✓ | ✓ | ✓ |
```
It was actually not so clear what the RFC means by that "other countermeasure options" as well. I would dig deeper into this so that we can come to a conclusion whether to merge this to 3.8.3 or just clarify the wording to make it clearer.

csfreak92 replied:
Upon reading further, it seems like it makes sense for me to combine **3.8.3** and **3.8.4** into one. I was worried before that it will be a very wordy and long requirement (which we know leads to some parts being omitted when people consider the verification requirement. I have a plan to make it a good one, stay tuned for the new commit.

csfreak92 commented on 5.0/en/0x12-V3-Session-management.md:
```diff
@@ -91,10 +90,88 @@ Attackers start a half-open attack by attempting to lock, reset, or recover a cr
+
+| # | Description | L1 | L2 | L3 |
+| :---: | :--- | :---: | :---: | :---: |
+| **3.8.5** | [ADDED] Verify that Clients are preventing injection (replay) of authorization codes into the authorization response by using PKCE flow. Additionally, clients may use the OpenID Connect "nonce" parameter and the respective Claim in the ID Token instead. The PKCE challenge or OpenID Connect "nonce" must be transaction-specific and securely bound to the client and the user agent in wh ...
```
I like it. That's way better wording that what I have. I'll adjust them accordingly to reflect this.

tghosth replied:
Great, thanks

tghosth commented:
@csfreak92, I went through your responses so far, keep at it :)

csfreak92 commented on 5.0/en/0x12-V3-Session-management.md:
```diff
@@ -91,10 +90,88 @@ Attackers start a half-open attack by attempting to lock, reset, or recover a cr
+| # | Description | L1 | L2 | L3 |
+| :---: | :--- | :---: | :---: | :---: |
+| **3.8.5** | [ADDED] Verify that Clients are preventing injection (replay) of authorization codes into the authorization response by using PKCE flow. Additionally, clients may use the OpenID Connect "nonce" parameter and the respective Claim in the ID Token instead. The PKCE challenge or OpenID Connect "nonce" must be transaction-specific and securely bound to the client and the user agent in wh ...
+| **3.8.6** | [ADDED] Verify that when using PKCE, Clients should use PKCE code challenge methods that do not expose the PKCE verifier in the authorization request. Otherwise, attackers that can read the authorization request can break the security provided by the PKCE. Authorization servers must support PKCE.  | ✓ | ✓ | ✓ |
```
Let me review the RFC again and some of its specifications as to why this was separate. I think it has to do with the phrase, "the PKCE code challenge methods do not expose the PKCE verifier in the authorization request" as the first thing I can think of this. I'll get back to you on this.

csfreak92 replied:
@tghosth, after reading a bunch, yup I agree regarding this **3.8.6** requirement it is redundant. I will remove it in my latest commit. The info from 3.8.6 text is quite useful so let me see if I can add it to the header of this subsection.

csfreak92 commented on 5.0/en/0x12-V3-Session-management.md:
```diff
@@ -91,10 +90,88 @@ Attackers start a half-open attack by attempting to lock, reset, or recover a cr
+* `Resource Owner (RO)` --- refers to an entity capable of granting access to a protected resource. When the resource owner is a person, it is referred to as an end-user. 
+* `Resource Server (RS)` --- refers to the server hosting the protected resources, capable of accepting and responding to protected resource requests using access tokens. 
+
+
```
Do you want something like a column or a marker that says who each specific requirement is for such as the developer of the Client, the developer of the Authorization Server, the developer of the Resource Server, etc?

tghosth replied:
Ideally I would like each to be in a different section like 3.8.x, 3.9.x, 3.10.x, etc.

csfreak92 replied:
Hmmm, let me see if I can devise that in a way to fit that preference. I grouped them according to the "topics"/OAuth use-cases.

csfreak92 replied:
@tghosth, here is an example of what I think can work as a header to distinguish the main implementor concerned for each of the new OAuth 2.0 requirements. What do you think?

> 
| # | Description | L1 | L2 | L3 | Implementor |
| :---: | :--- | :---: | :---: | :---: | :---: |  
| **3.8.1** | [ADDED] Verify that the Authorization Server supports CSRF protection either using the mechanism provided by PKCE, the "nonce" parameter or the "state" parameter. | ✓ | ✓ | ✓ | Authorization Server Developer |

tghosth replied:
I don't really want to add another column, I think I would prefer a section title.

csfreak92 replied:
Ok, let me work some magic on it by grouping things together.

csfreak92 replied:
Hi @tghosth, 

I'm having a hard time making your request for making it a per implementor-based list of requirements given the subsections of the proposed OAuth-related requirements. Another problem I have seen is there would be duplicate requirements if we will do it on an implementor-based list only such as in the following examples: 


## OAuth 2.0 Essential Basics

Some important Info text here. 

**Authorization Server Developer** 
| # | Description | L1 | L2 | L3 |
| :---: | :--- | :---: | :---: | :---: |
| **3.8.1** | [ADDED] Verify that the Authorization Server supports CSRF protection either using the mechanism provided by PKCE, the "nonce" parameter or the "state" parameter. | ✓ | ✓ | ✓ | 
| **3.8.4** | [ADDED] Verify that if a Client sends a valid PKCE "code_challenge" parameter in the authorization request, the Authorization Server enforces the correct usage of "code_verifier" at the token endpoint. | ✓ | ✓ | ✓ |
| **3.8.5** | [ADDED] Verify that Authorization Servers are mitigating PKCE Downgrade Attacks by ensuring a token request containing a "code_verifier" parameter is accepted only if a "code_challenge" parameter was present in the authorization request. | ✓ | ✓ | ✓ |


**Client Developer**
| # | Description | L1 | L2 | L3 |
| :---: | :--- | :---: | :---: | :---: |
| **3.8.2** | [ADDED] Verify that when an OAuth Client can interact with more than one Authorization Server, Clients should verify that the issuer "iss" parameter value is what it expected from the authorization response to prevent against mix-up attacks. [reference link for Mix-up attacks: https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics-18#mix_up] In the absence of "iss" parameter, Clients may instead use distinct redirect URIs to identify authorization endpoints and token endpoints.  | ✓ | ✓ | ✓ | 


## PKCE - Proof Key for Code Exchange Mechanism / Authorization Code Grant

Info text here about PKCE. 

**Authorization Server Developer** 
| # | Description | L1 | L2 | L3 |
| :---: | :--- | :---: | :---: | :---: |
| **3.8.3** | [ADDED] Verify that the replay of authorization codes into the authorization response is prevented either by using the PKCE flow or alternatively the OpenID Connect "nonce" parameter and the respective Claim in the ID Token. The PKCE challenge or OpenID Connect "nonce" must be transaction-specific and securely bound to the client and the user agent in which the transaction was started.  | ✓ | ✓ | ✓ |

**Client Developer**
| # | Description | L1 | L2 | L3 |
| :---: | :--- | :---: | :---: | :---: |
| **3.8.3** | [ADDED] Verify that the replay of authorization codes into the authorization response is prevented either by using the PKCE flow or alternatively the OpenID Connect "nonce" parameter and the respective Claim in the ID Token. The PKCE challenge or OpenID Connect "nonce" must be transaction-specific and securely bound to the client and the user agent in which the transaction was started.  | ✓ | ✓ | ✓ |


Requirement **3.8.3** has to be implemented both by the authorization server developer and the client developer. It is going to be duplicated if we do it this way as presented if we will do them by implementor. 

Also, notice that while **3.8.1** fits in the OAuth Essentials subsection, **3.8.4** and **3.8.5** falls under PKCE making it a little bit confusing to read while it is grouped well under Authorization Server Developer as the main implementor of those controls. 

Another solution I can think of is have all the subsections explained on top such as PKCE, Basics, Token Replay Prevention, Access Token Privilege Restriction, etc. and then after all of those texts were stated with their relevant info we list all the implementor-based controls list such as shown here, but it still doesn't solve the duplicates though. 

I would propose going back to having a suffixed "Implementor" column for each requirement to avoid duplicates and misplaced requirements/groupings. Let me know your thoughts.

tghosth replied:
### 3.8.3

If the Authorization Server (AS) implements this functionality, does the client have any choice whether to implement it or not? It feels like this is just a requirement on the AS.

Even if the client does have the choice whether to implement it or not, the requirement on the client is different because the client just needs to make sure it is using the functionality, not create the functionality.

### 3.8.4/3.8.5

I'm not sure I see the difference between these requirements

### 3.8.1

I'm not sure I understand the question. It seems like an AS requirement.


---

Overall, I really don't want to have another column here and I think we need to try and make this "who is the implementor" framework work.

csfreak92 replied:
Alright, I think I have a bit of an idea how to make it a little bit cleaner even if it seems messy. Stay tuned for a new commit next week for this PR!

csfreak92 commented on 5.0/en/0x12-V3-Session-management.md:
```diff
@@ -91,10 +90,88 @@ Attackers start a half-open attack by attempting to lock, reset, or recover a cr
+
+| # | Description | L1 | L2 | L3 |
+| :---: | :--- | :---: | :---: | :---: |
+| **3.8.9** | [ADDED] Verify that Clients are using the response type "code" (aka authorization code grant type) or any other response type that causes the Authorization Server to issue access tokens in the token response, such as the "code id_token" response type. This allows the Authorization Server to detect replay attempts by attackers and generally reduces the attack surface since access tok ...
```
Ok, will reword this one as well. Yeah, after reading it a few times it seems confusing to me now.

csfreak92 replied:
@tghosth, after hours of reading I realized you were right. I am going to remove this proposed **3.8.9** requirement as it talks about Implicit Flow for OAuth 2.0. Implicit Flow is pretty insecure, so it doesn't make sense to keep it unless you disagree and there is a way to somehow make it work? Let me know your thoughts on it.

tghosth replied:
Yeah let's leave it out for now unless we just want a requirement which says don't use it 🤣

Do we need to provide a secure way of doing it, e.g. using PKCE?

csfreak92 replied:
Yup, we have some requirements about PKCE. I'll remove this implicit flow then.

tghosth replied:
Please wait to resolve this conversation until you have made the relevant commit :)

csfreak92 commented on 5.0/en/0x12-V3-Session-management.md:
```diff
@@ -91,10 +90,88 @@ Attackers start a half-open attack by attempting to lock, reset, or recover a cr
+| :---: | :--- | :---: | :---: | :---: |
+| **3.8.12** | [ADDED] Verify that the privileges associated with an access token should be restricted to the minimum required for the particular application or use case. This prevents clients from exceeding the priviliges authorized by the Resource Owner. It also prevents users from exceeding their privileges authorized by the respective security policy. Privilege restrictions also help to reduc ...
+| **3.8.13** | [ADDED] Verify that access tokens are restricted to certain Resource Servers (audience restriction), preferably to a single Resource Server. Every Resource Server is obliged to verify, for every request, whether the access token sent with that request was meant to be used for that particular Resource Server. If not, the Resource Server must refuse to serve the respective request. C ...
+| **3.8.14** | [ADDED] Verify that access tokens are restricted to certain resources and actions on Resource Servers or resources. Every Resource Server is obliged to verify, for every request, whether the access token sent with that request was meant to be used for that particular action on the particular resource. If not, the Resource Server must refuse to serve the respective request. Clients  ...
```
Do you mean differentiate 3.8.14 from 3.8.13, right? If yes, then I would say 3.8.13 is more focused on the Resource Servers while 3.8.14 is more focused on the actions regarding resources. Moreover, 3.8.13 is requiring Clients to utilize the parameters "scope" and "resource" to determine the Resource Server they want to access while 3.8.14 is requiring Clients to utilize the parameters "scope" and "authorization_details" to determine the desired resources and/or actions. 

I know it seems very similar, and from what I read on the RFC it caught me off guard a few times, but separating them as even highlighted in the RFC seems logical and has some merit to it, in my opinion.

tghosth replied:
Apologies, I was specifically referring to differentiating between 3.8.14 and 3.8.12. Is 3.8.12 just referring to principle of least privilege because if so I think we should talk about that generically in chapter 4 access control but not here.

csfreak92 replied:
Yes, 3.8.12 specifically deals with principle of least privilege and access controls in relation to the Clients accessing resources. @tghosth, do you want me to add this Oauth requirement in chapter 4 access controls instead of keeping it together as bundled with the rest of the Oauth stuff?

tghosth replied:
I am pretty sure that [4.1.3](https://github.com/OWASP/ASVS/blob/master/5.0/en/0x12-V4-Access-Control.md#v41-general-access-control-design) already covers "principle of least privilege", no? Is there anything else that needs to be added?

csfreak92 commented:
@tghosth, @jmanico, @elarlang, @danielcuthbert, any idea how to identify which RFCs are already standardized and being followed widely? For context: I was reviewing the RFC for OAuth 2.0 regarding my ASVS pull request revisions and I noticed that there are newer versions of the RFC that I was reviewing which made me panick a little bit thinking I messed up a bunch of stuff, but turns out all of them have "Best Current Practice" statuses. I was wondering how will we know which is already turned into an acceptable standard by the IETF from the RFCs? I looked at each of the OAuth 2.0 RFCs and all of them only have "Best Current Practice" statuses. 

This also brings me to another question, should I stick with OAuth 2.0 RFC version 18 published in October 2021 or the latest version 23 to be published in December 2023? I noticed they have RFCs every few months, which makes these set of OAuth ASVS requirements quite a moving target if we are aiming to get the latest recommendations.

elarlang commented:
> Which are standardized

https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics-23

I think it gets RFC number when it is "released".

Compare with https://datatracker.ietf.org/doc/html/6265

> and actively followed

I'm not sure it is possible or easy to identify that.


"Best Current Practices" - I watch it as Living standard like we have for HTML and URL nowadays. If something makes sense, we put it to ASVS as requirement. If something gets updated, then there should be good reason behind that and we can not ignore that.

In short - I can not see any reason to stick with old version, when there is new available. Yes, it requires more work, but we need to check, what and why they changed. I have not checked the diff, maybe there are no meaningful changes from ASVS requirements point of view at all.

It's also points out why I don't like linking requirement to RFC's too much, because we probably fix those requirement for some 4 years (ASVS v4.0 is 4+ years old now) and we can not fix requirements (those could mean "breaking changes"). But for v5.0 we have some requirements pointing to RFC's anyway.

csfreak92 commented:
Alright, I'll do the modifications from Josh first then look at the diff from version 18 and 23. I think I saw a few new things added, so if those make sense I would add them as well. 

@elarlang, can you expand a little bit about what you mean by this? Do you mean some of the RFC's requirements are mostly covered in older ASVS versions? I got confused with the breaking changes you mentioned. 

> It's also points out why I don't like linking requirement to RFC's too much, because we probably fix those requirement for some 4 years (ASVS v4.0 is 4+ years old now) and we can not fix requirements (those could mean "breaking changes"). But for v5.0 we have some requirements pointing to RFC's anyway.

csfreak92 commented:
Ah, I see now what you mean... yeah that is quite another problem in case there are breaking changes. I'll get through the revisions, then check out the newly added diffs in the newer RFCs, hopefully no breaking changes.

csfreak92 commented:
@tghosth, I have pushed a latest commit on this branch for the OAuth 2.0 requirements. I have cleaned them up, added one or two from the latest RFC requirements and then grouped the proposed requirements based on the implementors needed. Please take note that the 3.8.xx numbers have changed as I have adjusted them. 

There is a "Mapping" note that I have added, let me know if you wish to still have this or shall I remove it? It seems important and useful info, but let me know your thoughts.

csfreak92 commented:
Ping @tghosth for review please.
